prov/gni/src/gnix_mr.c

```diff
@@ -295,14 +360,101 @@ DIRECT_FN int gnix_mr_regattr(struct fid *fid, const struct fi_mr_attr *attr,
 		return -FI_EOPNOTSUPP;
 
 	if (attr->iov_count == 1)
-		return __mr_reg(fid, attr->mr_iov[0].iov_base, 
+		return _gnix_mr_reg(fid, attr->mr_iov[0].iov_base, 
 			attr->mr_iov[0].iov_len, attr->access, attr->offset,
-			attr->requested_key, flags, mr, attr->context);
+			attr->requested_key, flags, mr, attr->context, 
+			USER_REGISTRATION);
 	
 	/* regv limited to one iov at this time */
 	return -FI_EOPNOTSUPP;
 }
 
+static int __gnix_mr_refresh(struct gnix_fid_mem_desc *desc, uint64_t addr, uint64_t len)
+{
+		gni_return_t grc;
+
+
+	fastlock_acquire(&desc->nic->lock);
+	GNIX_INFO(FI_LOG_MR, "hndl=%p addr=%p len=%d cq=%p flags=%x vmdh_index=%d mem_hndl=%p\n",
+		desc->nic->gni_nic_hndl, addr, len, NULL, GNI_MEM_UPDATE_REGION,
+		desc->mr_fid.key, &desc->mem_hndl);
+	grc = GNI_MemRegister(desc->nic->gni_nic_hndl, addr,
+		len, NULL, GNI_MEM_UPDATE_REGION,
+		desc->mr_fid.key, &desc->mem_hndl);
+	assert(grc == FI_SUCCESS); /* TODO: handle error case */
```

Comment:
Ok, I would be worried about silent failures. Maybe for testing if you're not compiling with --enable-debug you could do something like:
```
#ifndef ENABLE_DEBUG
#define assert(cond) (if ((cond)) {printf("assert failed at %d\n", __LINE__); abort();})
#endif
```